packages/shared/src/contexts/UITaskManager.tsx

```diff
@@ -0,0 +1,99 @@
+import { defer, DeferTuple } from '@dimensiondev/kit'
+import { EMPTY_LIST } from '@masknet/shared-base'
+import { ComponentType, createContext, createElement, FC, PropsWithChildren, useMemo, useState } from 'react'
+import type { InjectedDialogProps } from './components'
+
+type ExtendPromise<T> = Promise<T> & {
+    close(): Promise<T>
+}
+interface ContextOptionsGeneric<T, R> {
+    show(options: T): ExtendPromise<R>
+}
+
+type BaseDialogProps = Pick<InjectedDialogProps, 'open' | 'onClose'>
+type CallableKeys<T> = keyof {
+    [K in keyof T as T[K] extends ((...args: any[]) => any) | undefined ? K : never]: any
+}
+
+/**
+ * Create a manager of small UI task sessions,
+ * which provide both a Context and a Provider.
+ */
+export const createUITaskManager = <
+    TaskOptions extends Record<string, any>,
+    Result,
+    Props extends BaseDialogProps,
+    ResolveProp extends CallableKeys<Props> & CallableKeys<TaskOptions>,
+>(
+    Component: ComponentType<Props>,
+    taskResolveField: ResolveProp,
+) => {
+    type ContextOptions = ContextOptionsGeneric<TaskOptions, Result | null>
+    const TaskManagerContext = createContext<ContextOptions>(null!)
+    let id = 0
+
+    type TaskDeferTuple = DeferTuple<Result | null>
+    interface TaskBase {
+        id: number
+        promise: Promise<Result | null>
+        resolve: TaskDeferTuple[1]
+        reject: TaskDeferTuple[2]
+    }
+    interface Task extends TaskBase {
+        taskOptions: TaskOptions
+    }
+
+    const TaskManagerProvider: FC<PropsWithChildren<{}>> = ({ children }) => {
+        const [tasks, setTasks] = useState<Task[]>(EMPTY_LIST)
+
+        const contextValue = useMemo(() => {
+            const removeTask = (id: number) => {
+                setTasks((list) => list.filter((t) => t.id !== id))
+            }
+
+            return {
+                show(options: TaskOptions) {
+                    const [promise, resolve, reject] = defer<Result | null>()
+                    id += 1
+                    const extendedPromise = Object.assign(promise, {
+                        close() {
+                            resolve(null)
```

Comment:
You should use reject instead of resolve null.

Reply:
No, aborting should not raise an exception.
```ts
const pickeToken = await pickToken()
// get null by just closing the dialog
// like `window.prompt`
const text = prompt('what's your name?')
```

Reply:
Then what if `Result` contains `null`? You cannot distinguish between these two different states because you express them both in `null`.

Reply:
>Then what if Result contains null?

That means "I don't care about the result at all"

Reply:
You should distinguish result/close with one of the following way:

- Promise rejection
- `Some<T>`
- Tagged union (`{ type: 'close' } | { type: 'resolve', value: T }`)

Reply:
I prefer to keep it simple.
We still have chance to do such refactor if we need it.